Create @frontity/head-tags package

DAreRodz opened the pull request:
<!--
Thanks for your pull request 😊. Note that not following the template might result in your issue being closed
-->

#### Description of what you did:

<!-- Please describe everything as much as possible -->

This PR adds a new package called `@frontity/head-tags` that integrates with **REST API - Head Tags by Frontity** WordPress plugin and insert head tags for each page using the `<Head>` component.

The URLs pointing to WordPress blog pages (i.e. pages that not contain `wp-json`,  `wp-content`, `wp-admin`, etc.), would be converted to Frontity URLs.

#### My PR is a:

<!-- Delete the ones that don't apply -->

- 🚀 New feature

#### Is the PR ready to be merged?

<!-- Delete the one that don't apply -->

- ✅ Yes!
<!-- - 🚧 Work in progress -->

DAreRodz committed:
chore(head-tags): create head-tags package

DAreRodz committed:
chore(head-tags): fix type of tag prop

DAreRodz committed:
chore(head-tags): add comments and some fixes

DAreRodz committed:
chore(head-tags): fix head_tags field name

DAreRodz committed:
chore(source): add a type for post types

DAreRodz committed:
chore(head-tags): improve typing

DAreRodz committed:
feat(wp-source): add post types to state

DAreRodz committed:
chore(examples): revert changes in mars settings

DAreRodz committed:
chore(npm): update dependencies

DAreRodz committed:
refactor(source): refactor entity types

DAreRodz committed:
refactor(wp-source): change schemas and state

Add postType and taxonomyType.
Add `Entity` to names of old schemas.
Add `type` and `taxonomy` props to `source.state`.
Update tests.

DAreRodz committed:
chore(head-tags): add wp-source as dependency

DAreRodz committed:
chore(head-tags): transform links to blog pages

DAreRodz committed:
chore(head-tags): move a type definition

DAreRodz committed:
test(head-tags): add test.todo

DAreRodz committed:
Fix postType schema and add tests

DAreRodz committed:
Improve source types

DAreRodz committed:
Add test for post type

DAreRodz committed:
Move functions from component to utils

DAreRodz committed:
Fix getCurrentHeadTags [WIP]

DAreRodz committed:
Fix getCurrentHeadTags

DAreRodz committed:
Change also URLs inside ld+json content

DAreRodz committed:
Move current head tags to a derived value

DAreRodz committed:
Change all test by more specific ones

DAreRodz committed:
Finish tests for post entities

DAreRodz committed:
Add tests for post types

DAreRodz committed:
Merge branch 'dev' into head-tags

DAreRodz committed:
Change test suite names

DAreRodz committed:
Add tests for taxonomies and authors

DAreRodz committed:
Add tests for links with no entity

DAreRodz committed:
Set initial version to 0.1.0

DAreRodz committed:
CHANGESET: silent-bees-prove. Release the first version of @frontity/head-tags package.

DAreRodz committed:
CHANGESET: fifty-hairs-explode. Change the name of entity types:

DAreRodz committed:
CHANGESET: serious-bears-sniff. Add a schema for post types and refactor populate.

DAreRodz committed:
Update changeset descriptions

luisherranz commented on packages/head-tags/package.json:
```diff
@@ -0,0 +1,42 @@
+  "license": "Apache-2.0",
+  "keywords": [
+    "frontity",
+    "frontity-yoast",
```
We have to use here `"frontity-head-tags"`.

luisherranz commented on packages/head-tags/package.json:
```diff
@@ -0,0 +1,42 @@
+    "typescript": "^3.6.3"
+  },
+  "dependencies": {
+    "string-humanize": "^1.0.1"
```
This can be deleted.

luisherranz commented on packages/head-tags/types.ts:
```diff
@@ -0,0 +1,58 @@
+  AuthorEntity,
+  PostType
+} from "@frontity/source";
+import { Merge } from "@frontity/source/src/utils";
```
We can move `Merge` to this file. It's safer because now this package can break if someone decides to refactor the utils of `@frontity/source`.

luisherranz commented on packages/head-tags/src/utils/index.ts:
```diff
@@ -0,0 +1,153 @@
+  const headTags = (entity && entity.head_tags) || [];
+
+  // Changes those links that points to WordPress blog pages to Frontity links
+  return useFrontityLinks({ state, headTags });
```
If `entity.head_tags` is null, we don't need to do `useFrontityLinks`.

luisherranz commented on packages/head-tags/src/utils/index.ts:
```diff
@@ -0,0 +1,153 @@
+  headTags: HeadTags;
+}) => {
+  // The site URL.
+  const frontityUrl = state.frontity.url;
```
Maybe we can stop here if this doesn't exist with a warning in the console.

It could be a good moment to add the `@frontity/error` package.

luisherranz replied:
What about using a default export with a function, like:
```js
import msg from "@frontity/error";

console.warn(msg("Something is not quite right"));
throw new Error(msg("Something went really bad"));
```

luisherranz commented on packages/head-tags/types.ts:
```diff
@@ -0,0 +1,58 @@
+      url: string;
+    };
+    source?: Merge<
+      WpSource["state"]["source"],
```
I think we should use `Source` here, and add `isWpCom` as optional:
```ts
source?: Merge<
      Source["state"]["source"],
      {
        isWpCom?: boolean;
        post: Record<string, PostEntityWithHeadTags>;
        page: Record<string, PostEntityWithHeadTags>;
        author: Record<string, AuthorEntityWithHeadTags>;
        type: Record<string, PostTypeWithHeadTags>;
        category: Record<string, TaxonomyWithHeadTags>;
        tag: Record<string, TaxonomyWithHeadTags>;
      }
    >
```

luisherranz commented on packages/head-tags/src/utils/index.ts:
```diff
@@ -0,0 +1,153 @@
+    return new URL(pathname.replace(/^\/wp\/v2\/sites\//, "https://"));
+  }
+  // Get API subdirectory.
+  const apiSubdir = apiUrl.pathname.replace(/\/wp-json\/?$/, "/");
```
Can we use `state.source.subdirectory` here to avoid hardcoding it to WP?

DAreRodz replied:
I think we misunderstood `state.source.subdirectory` the other day, because it's the Frontity site's subdirectory and not the WordPress site's, so we would have to add a new setting for that.

luisherranz commented on packages/head-tags/src/index.ts:
```diff
@@ -0,0 +1,15 @@
+  roots: { headTags: Component },
+  state: {
+    headTags: {
+      current: ({ state }) => getCurrentHeadTags({ state })
```
I will explore the possibility of using a more general API, like `state.headTags.get(link)`.

luisherranz commented on packages/head-tags/src/__tests__/state.tests.ts:
```diff
@@ -0,0 +1,540 @@
+    // Populate all state.
+    setUpState(store.state, headTags);
+    // Test current head tags.
+    expect(store.state.headTags.current).toMatchSnapshot();
```
I would use `toEqual([])` here to be more explicit in what it expects.

luisherranz commented on packages/head-tags/src/components/index.tsx:
```diff
@@ -0,0 +1,23 @@
+const Root: React.FC<Connect<HeadTagsPackage>> = ({ state }) => {
+  return (
+    <Head>
+      {state.headTags.current.map(
```
You can use `useMemo` for this computation (https://reactjs.org/docs/hooks-reference.html#usememo).

luisherranz commented on packages/source/src/entities.ts:
```diff
@@ -7,7 +10,15 @@ interface BaseEntity {
   _links?: any;
```
I think it's safer to use `unkwown` than `any`.

luisherranz commented on .changeset/serious-bears-sniff.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@frontity/wp-source": patch
```
This should be a minor.

luisherranz reviewed (changes requested):
Great work @DAreRodz. I've added a few comments in the code but looks great :)

DAreRodz committed:
Fix keywords and remove unused dependencies

DAreRodz committed:
Copy Merge type and change WpSource type by Source

DAreRodz committed:
Don't run useFrontityLinks if no entity/head tags

DAreRodz committed:
Change state.current by state.get(link)

DAreRodz committed:
Optimize headTags.get(link) with React.useMemo

DAreRodz committed:
Change all `any` types to `unknown`

DAreRodz committed:
Set `wp-source` changeset to `minor`

DAreRodz committed:
Create @frontity/error package

DAreRodz committed:
CHANGESET: tidy-starfishes-serve. Create `@frontity/error` package.

DAreRodz committed:
Move `name` to the FrontityError's prototype

DAreRodz committed:
Warn if `state.frontity.url` is not defined

DAreRodz committed:
Replace `error` package by `message` package

DAreRodz committed:
Refactor head-tags package to be more hackable

DAreRodz committed:
Update dependencies

DAreRodz committed:
Add test for headTags options

DAreRodz committed:
Replace error package by message one in changesets

DAreRodz committed:
Change `prefix` to `base`

DAreRodz committed:
Respect trailing slash

luisherranz reviewed (approved):
Awesome! Looks great :)

Let's remove `@frontity/message` package and add it to the roadmap to think more about the different use cases.

DAreRodz committed:
Add more dependencies to React.useMemo

DAreRodz committed:
Delete `@frontity/message` package

DAreRodz committed:
Add a warning for invalid ld+json content

luisherranz committed:
Merge branch 'dev' into head-tags

luisherranz committed:
Add engines to package.json

This is needed to update the node version in Now because they have
discontinued node 8. I think it can be removed later.
